README.md
# Summary
Interactive shell for the [WinUAE](http://www.winuae.net/) and [FS-UAE](https://fs-uae.net/) Amiga emulators.

# Installation
To install the package:

> `npm install -g amishell-1.0.0.tgz`

To verify that the command is correctly installed:

> `amishell --help`

To uninstall the package:

> `npm uninstall -g amishell`

# Setup
The Amiga emulator needs to be properly configured for communication through a virtual serial port.

## FS-UAE  
Include the following property in your .fs-uae configuration file:

```
serial_port = tcp://0.0.0.0:1234`
```

## WinUAE
Include the following lines in your .uae configuration file:

```
win32.serial_port=TCP://0.0.0.0:1234 serial_translate=crlf_cr
serial_direct=true
```

## User-Startup
Add the following lines to your Amiga `S:User-Startup` file:

```
mount AUX:
newshell AUX:
```

You will need to have the `AUX` mountlist in the `DEVS:DosDrivers` directory.

You may want to rename the `AUX` mountlist as `AUXI` under Windows if you have any problems with that particular name.

## Serial Preferences
Execute the `Prefs:Serial` (2.0+) or `Prefs:Preferences` (1.3) program in your Amiga hard drive and set the serial preferences to these values:

* Baud Rate: 31250
* Input Buffer Size: 512
* Handshaking: XON/XOFF
* Parity: None
* Bits/Char: 8
* Stop Bits: 1

# Usage
1. Open your properly-configured Amiga emulator of choice.
2. Open a Terminal or Command Prompt window.
3. To execute a single command:
	> `amishell dir`
4. To start the interactive shell:
	> `amishell`
5. Use Up/Down arrows for command history.
6. Use Ctrl+C to exit.

# Credits
* Coding: Jose Moreno 'Xeleh' (<jmoreno@xeleh.com>).
* Testing: Javier Romero 'Tolkien'.


CHANGELOG.md
## v1.0.0

- Initial version.


src/amishell.ts
import * as cp from "child_process";
import * as net from "net";
import { EventEmitter } from "events";

export class AmiShell extends EventEmitter {

private emulator: string;
private port: number;
private scripts: string;

constructor(emulator: string, port: number, scripts: string = null) {
	super();
	this.emulator = emulator;
	this.port = port;
	if (scripts == null) {
		scripts = __dirname + "/../scripts";
	}
	this.scripts = scripts;
}

public async executeCommand(command: string, activate: boolean, timeout: number, 
  quiet: boolean = false): Promise<string> {
	if (activate) {
		this.activateEmulator();
		await this.sleep(250);
	}
	await this.createSocket();
	return await this.sendCommand(command, timeout, quiet);
}

private sleep(ms: number) {
    return new Promise(resolve => {
        setTimeout(resolve, ms);
    });
}

private activateEmulator() {
	let command = "";
	switch (process.platform) {
	case "darwin":
	case "linux":
		command = "\"" + this.scripts + "/activate.sh" + "\" " + "fs-uae";
		break;
	case "win32":
		let winPath = this.scripts.replace("/", "\\");
		command = "cscript //nologo \"" + winPath + "\\activate.vbs" + "\" ";
		switch (this.emulator) {
		case "fsuae":
			command += "fs-uae.exe";
			break;
		case "winuae":
			command += "winuae.exe winuae64.exe";
			break;
		}
		break;
	}
	if (command != "") {
		cp.exec(command, (error, stdout, stderr) => {
			if (error) {
				this.destroySocket();
				this.emit("error", error);
			}
		});
	}
}

private socket: net.Socket;

private async createSocket() {
	return new Promise(resolve => {
		this.socket = net.createConnection(this.port, "localhost", () => { 
			this.socket.setEncoding("ascii");
			resolve();
		});
		this.socket.on("error", (error) => {
			this.destroySocket();
			this.emit("error", error);
		});
	});
}

public eot : boolean = false;

private async sendCommand(command: string, timeout: number, quiet: boolean): Promise<string> {
	return new Promise<string>(resolve => {
		let result : string = "";
		let dataLength = 0;
		this.eot = false;
		let prompt = "";
		this.socket.on("data", (data) => {
			var response = data.toString();
			// AmigaDos sends ascii char SI (15) as EOT signal
			if (response.charCodeAt(0) == 15) {
				this.eot = true;
			}
			if (!this.eot) {
				response = response.replace("\r", "");
				// filter the command itself
				if (dataLength > command.length) {
					if (!quiet) {
						this.emit("data", response);
					}
					result += response;
				}
				dataLength += response.length;
			} else {
				// wait for complete prompt
				prompt += response;
				if (prompt.lastIndexOf("> ") >= 0) {
					this.socket.setTimeout(10);
				}
			}
		});
		// the promise will be resolved on timeout
		this.socket.setTimeout(timeout);
		this.socket.on("timeout", () => {
			this.destroySocket();
			resolve(result);
		});
		// send the command
		this.socket.write(command + "\r");
	});
}

private destroySocket() {
	if (this.socket && !this.socket.destroyed) {
		this.socket.end();
		this.socket.destroy();
	}
}

}


src/main.ts
#!/usr/bin/env node

import { AmiShell } from "./amishell";
import * as readline from "readline";

const versionLabel = "1.0";

var amishell : AmiShell;

parseArgs();

async function parseArgs() {
	let emulator = (process.platform === "win32") ? "winuae" : "fsuae";
	const optionDefinitions = [
	  { name: 'activate', alias: 'a', type: Boolean, defaultValue: false },
	  { name: 'emulator', alias: 'e', type: String, defaultValue: emulator },
	  { name: 'help', alias: 'h', type: Boolean },
	  { name: 'port', alias: 'p', type: Number, defaultValue: 1234 },
	  { name: 'timeout', alias: 't', type: Number, defaultValue: 500 },
	  { name: 'version', alias: 'v', type: Boolean },
	  { name: 'command', type: String, multiple: true, defaultOption: true }
	];
	const cliArgs = require('command-line-args');
	const args = cliArgs(optionDefinitions, { partial: true });
	if (args._unknown && args._unknown.length > 0) {
		console.log("amishell: unrecognized option '" + args._unknown[0] + "'");
		process.exit(1);
	}
	if (args.help) {
		help();
	}
	if (args.version) {
		version();
	}
	amishell = new AmiShell(args.emulator, args.port);
	amishell.on("data",	(data) => { 
		process.stdout.write(data);
	});
	amishell.on("error", (error) => {
		console.error("[Communication with Amiga emulator failed]");
		console.error(error.message);
		terminate(1);
	});
	if (args.command) {
		let command = args.command.join(" ");
		await amishell.executeCommand(command, args.activate, args.timeout);
		process.exit(0);
	}
	shell(args.activate, args.timeout);
}

function help() {
	console.log("Executes commands in a running Amiga emulator.")
	console.log("Usage: [options] <command>");
	console.log("\nOptions");
	console.log("-a, --activate             Activate (send to front) the emulator window");
	console.log("-e, --emulator <emulator>  Emulator in use (fsuae or winuae)");
	console.log("-h, --help                 Show this help");
	console.log("-p, --port     <port>      Port used by the emulator for serial comm. (default = 1234)");
	console.log("-t, --timeout  <timeout>   Timeout in milliseconds (default = 500)");
	console.log("-v, --version              Show the version number");
	console.log("\nSpecify no <command> to enter Shell mode.");
	process.exit(0);
}

function version() {
	console.log("amishell v" + versionLabel);
	process.exit(0);
}

async function shell(activate: boolean, timeout: number) {
	let rl = readline.createInterface({
		input: process.stdin,
		output: process.stdout
	});
	let busy = false;
	await prompt(rl);
	rl.on('line', async (command: any) => {
		if (!busy) {
			busy = true;
			await amishell.executeCommand("", false, 100, true);
			if (amishell.eot) {
				await amishell.executeCommand(command, activate, timeout);
			}
			await prompt(rl);
			busy = false;
		}
	});
	rl.on('close', () => {
		console.log('\n[Terminated by user]');
		terminate();
	});
}

async function prompt(rl: readline.ReadLine) {
	let cwd = await amishell.executeCommand("cd", false, 250, true);
	cwd = amishell.eot ? cwd.replace("\n","") + "> " : "";
	rl.setPrompt(cwd);
	rl.prompt();
}

function terminate(code: number = 0) {
	process.exit(code);
}
